File: apps/api/app/services/abstract_bust_blueprint_v2.py
from __future__ import annotations

import json
import math
from copy import deepcopy
from enum import Enum
from pathlib import Path
from typing import Any, Dict, Literal

from pydantic import BaseModel, ConfigDict


ABSTRACT_BUST_V2_GENERATOR_VERSION = "abstract_bust_v0_2"
ABSTRACT_BUST_V2_CONTRACT_ROOT = (
    Path(__file__).resolve().parents[4]
    / "packages"
    / "shared-schema"
    / "contracts"
    / ABSTRACT_BUST_V2_GENERATOR_VERSION
)
ABSTRACT_BUST_V2_SCHEMA_PATH = (
    ABSTRACT_BUST_V2_CONTRACT_ROOT / "abstract_bust_blueprint_v0_2.schema.json"
)
ABSTRACT_BUST_V2_DEFAULT_FIXTURE_PATH = (
    ABSTRACT_BUST_V2_CONTRACT_ROOT / "fixtures" / "default.json"
)


class AbstractBustBlueprintV2ValidationError(ValueError):
    def __init__(self, code: str, path: str, message: str):
        self.code = code
        self.path = path
        super().__init__(f"{code} at {path}: {message}")


def _reject_json_constant(value: str) -> None:
    raise AbstractBustBlueprintV2ValidationError(
        "non_finite_number", "$", f"{value} is not valid JSON"
    )


def _load_json(path: Path) -> Dict[str, Any]:
    value = json.loads(
        path.read_text(encoding="utf-8"),
        parse_constant=_reject_json_constant,
    )
    if not isinstance(value, dict):
        raise AbstractBustBlueprintV2ValidationError(
            "invalid_root_type", "$", "expected an object"
        )
    return value


_SCHEMA = _load_json(ABSTRACT_BUST_V2_SCHEMA_PATH)
_DEFAULT_BLUEPRINT = _load_json(ABSTRACT_BUST_V2_DEFAULT_FIXTURE_PATH)


def _schema_enum(path: tuple[str, ...]) -> list[str]:
    value: Any = _SCHEMA
    for component in path:
        value = value[component]
    if not isinstance(value, list) or not all(
        isinstance(candidate, str) for candidate in value
    ):
        raise RuntimeError(f"Invalid v0.2 enum schema at {'.'.join(path)}")
    return value


AbstractBustPresentationV2 = Enum(
    "AbstractBustPresentationV2",
    {
        value: value
        for value in _schema_enum(("properties", "presentation", "enum"))
    },
    type=str,
)
AbstractBustAgeTendencyV2 = Enum(
    "AbstractBustAgeTendencyV2",
    {
        value: value
        for value in _schema_enum(("properties", "age_tendency", "enum"))
    },
    type=str,
)
AbstractBustHairStyleV2 = Enum(
    "AbstractBustHairStyleV2",
    {
        value: value
        for value in _schema_enum(
            ("$defs", "hair", "properties", "style", "enum")
        )
    },
    type=str,
)


class AbstractBustModelV2(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)


class AbstractBustHeadV2(AbstractBustModelV2):
    width: float
    height: float
    depth: float
    roundness: float


class AbstractBustEyesV2(AbstractBustModelV2):
    vertical_position: float
    spacing: float
    size: float
    tilt: float
    contour_strength: float


class AbstractBustNoseV2(AbstractBustModelV2):
    vertical_position: float
    width: float
    length: float
    prominence: float


class AbstractBustMouthV2(AbstractBustModelV2):
    vertical_position: float
    width: float
    curvature: float
    contour_strength: float


class AbstractBustCheeksV2(AbstractBustModelV2):
    width: float
    vertical_position: float
    prominence: float


class AbstractBustJawV2(AbstractBustModelV2):
    width: float
    taper: float
    length: float
    roundness: float


class AbstractBustFaceV2(AbstractBustModelV2):
    eyes: AbstractBustEyesV2
    nose: AbstractBustNoseV2
    mouth: AbstractBustMouthV2
    cheeks: AbstractBustCheeksV2
    jaw: AbstractBustJawV2


class AbstractBustNeckV2(AbstractBustModelV2):
    width: float
    length: float


class AbstractBustShouldersV2(AbstractBustModelV2):
    width: float
    slope: float


class AbstractBustTorsoV2(AbstractBustModelV2):
    width: float
    thickness: float
    length: float
    taper: float


class AbstractBustHairV2(AbstractBustModelV2):
    style: AbstractBustHairStyleV2
    volume: float
    length: float


class AbstractBustContourV2(AbstractBustModelV2):
    softening: float
    asymmetry: float


class AbstractBustBlueprintV2(AbstractBustModelV2):
    generator_version: Literal["abstract_bust_v0_2"]
    particle_count: int
    presentation: AbstractBustPresentationV2
    age_tendency: AbstractBustAgeTendencyV2
    seed: int
    head: AbstractBustHeadV2
    face: AbstractBustFaceV2
    neck: AbstractBustNeckV2
    shoulders: AbstractBustShouldersV2
    torso: AbstractBustTorsoV2
    hair: AbstractBustHairV2
    contour: AbstractBustContourV2


def _resolve_schema(schema: Dict[str, Any]) -> Dict[str, Any]:
    reference = schema.get("$ref")
    if reference is None:
        return schema
    prefix = "#/$defs/"
    if not isinstance(reference, str) or not reference.startswith(prefix):
        raise RuntimeError(
            f"Unsupported AbstractBustBlueprintV2 schema reference: {reference!r}"
        )
    definition = _SCHEMA.get("$defs", {}).get(reference[len(prefix) :])
    if not isinstance(definition, dict):
        raise RuntimeError(
            f"Missing AbstractBustBlueprintV2 schema definition: {reference}"
        )
    return definition


def _validation_error(code: str, path: str, message: str) -> None:
    raise AbstractBustBlueprintV2ValidationError(code, path, message)


def _normalize_value(
    value: Any,
    schema: Dict[str, Any],
    default_value: Any,
    path: str,
) -> Any:
    schema = _resolve_schema(schema)

    if "const" in schema and (
        type(value) is not type(schema["const"]) or value != schema["const"]
    ):
        code = (
            "unknown_generator_version"
            if path == "$.generator_version"
            else "invalid_const"
        )
        _validation_error(code, path, f"expected {schema['const']!r}")

    enum_values = schema.get("enum")
    if isinstance(enum_values, list) and not any(
        type(value) is type(candidate) and value == candidate
        for candidate in enum_values
    ):
        _validation_error("unknown_enum", path, f"unsupported value {value!r}")

    schema_type = schema.get("type")
    if schema_type == "object":
        if not isinstance(value, dict):
            _validation_error("invalid_type", path, "expected object")
        properties = schema.get("properties")
        if not isinstance(properties, dict):
            raise RuntimeError(f"Object schema has no properties at {path}")
        unknown_fields = sorted(set(value) - set(properties))
        if unknown_fields:
            _validation_error(
                "unknown_field",
                f"{path}.{unknown_fields[0]}",
                "additional properties are forbidden",
            )
        required = schema.get("required")
        required_fields = set(required) if isinstance(required, list) else set()
        missing_required = sorted(
            field_name
            for field_name in required_fields
            if field_name not in value
        )
        if missing_required:
            _validation_error(
                "missing_required_field",
                f"{path}.{missing_required[0]}",
                "required field is missing",
            )

        defaults = default_value if isinstance(default_value, dict) else {}
        normalized: Dict[str, Any] = {}
        for field_name, field_schema in properties.items():
            if not isinstance(field_schema, dict):
                raise RuntimeError(f"Invalid schema for {path}.{field_name}")
            if field_name in value:
                normalized[field_name] = _normalize_value(
                    value[field_name],
                    field_schema,
                    defaults.get(field_name),
                    f"{path}.{field_name}",
                )
                continue
            if field_name in required_fields:
                continue
            if field_name in defaults:
                optional_default = deepcopy(defaults[field_name])
            elif "default" in field_schema:
                optional_default = deepcopy(field_schema["default"])
            else:
                continue
            normalized[field_name] = _normalize_value(
                optional_default,
                field_schema,
                optional_default,
                f"{path}.{field_name}",
            )
        return normalized

    if schema_type == "integer":
        if type(value) is not int:
            _validation_error("invalid_type", path, "expected integer")
        minimum = schema.get("minimum")
        maximum = schema.get("maximum")
        if isinstance(minimum, int):
            value = max(minimum, value)
        if isinstance(maximum, int):
            value = min(maximum, value)
        return value

    if schema_type == "number":
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            _validation_error("invalid_type", path, "expected number")
        numeric_value = float(value)
        if not math.isfinite(numeric_value):
            _validation_error(
                "non_finite_number", path, "number must be finite"
            )
        minimum = schema.get("minimum")
        maximum = schema.get("maximum")
        if isinstance(minimum, (int, float)):
            numeric_value = max(float(minimum), numeric_value)
        if isinstance(maximum, (int, float)):
            numeric_value = min(float(maximum), numeric_value)
        return numeric_value

    return value


def _strict_model_from_normalized(
    normalized: Dict[str, Any],
) -> AbstractBustBlueprintV2:
    strict_model_input = deepcopy(normalized)
    strict_model_input["presentation"] = AbstractBustPresentationV2(
        strict_model_input["presentation"]
    )
    strict_model_input["age_tendency"] = AbstractBustAgeTendencyV2(
        strict_model_input["age_tendency"]
    )
    strict_model_input["hair"]["style"] = AbstractBustHairStyleV2(
        strict_model_input["hair"]["style"]
    )
    return AbstractBustBlueprintV2.model_validate(strict_model_input)


def normalize_abstract_bust_blueprint_v2(value: Any) -> Dict[str, Any]:
    if isinstance(value, (str, bytes, bytearray)):
        try:
            value = json.loads(value, parse_constant=_reject_json_constant)
        except AbstractBustBlueprintV2ValidationError:
            raise
        except (TypeError, json.JSONDecodeError) as error:
            raise AbstractBustBlueprintV2ValidationError(
                "invalid_json", "$", str(error)
            ) from error
    normalized = _normalize_value(value, _SCHEMA, _DEFAULT_BLUEPRINT, "$")
    return _strict_model_from_normalized(normalized).model_dump(mode="json")


def parse_abstract_bust_blueprint_v2(value: Any) -> AbstractBustBlueprintV2:
    return _strict_model_from_normalized(
        normalize_abstract_bust_blueprint_v2(value)
    )


def load_default_abstract_bust_blueprint_v2() -> AbstractBustBlueprintV2:
    return parse_abstract_bust_blueprint_v2(deepcopy(_DEFAULT_BLUEPRINT))


def default_abstract_bust_blueprint_v2_dict() -> Dict[str, Any]:
    return load_default_abstract_bust_blueprint_v2().model_dump(mode="json")


_PRESET_FILENAMES = {
    "presentation": {
        "neutral": "neutral.json",
        "feminine": "feminine.json",
        "masculine": "masculine.json",
    },
    "age_tendency": {
        "youthful": "youthful.json",
        "balanced": "balanced.json",
        "mature": "mature.json",
    },
    "hair": {
        "none": "hair_none.json",
        "short": "hair_short.json",
        "medium": "hair_medium.json",
        "long": "hair_long.json",
        "tied": "hair_tied.json",
    },
}


def load_abstract_bust_preset_fixture_v2(
    kind: str,
    value: str,
) -> AbstractBustBlueprintV2:
    filename = _PRESET_FILENAMES.get(kind, {}).get(value)
    if filename is None:
        _validation_error(
            "unknown_enum",
            f"$.{kind}",
            f"unsupported value {value!r}",
        )
    return parse_abstract_bust_blueprint_v2(
        _load_json(ABSTRACT_BUST_V2_CONTRACT_ROOT / "fixtures" / filename)
    )


def apply_abstract_bust_preset_v2(
    current: Any,
    kind: str,
    value: str,
) -> AbstractBustBlueprintV2:
    normalized_current = normalize_abstract_bust_blueprint_v2(current)
    preset = load_abstract_bust_preset_fixture_v2(kind, value).model_dump(
        mode="json"
    )
    preset["seed"] = normalized_current["seed"]
    preset["particle_count"] = normalized_current["particle_count"]
    return parse_abstract_bust_blueprint_v2(preset)

File: packages/shared-schema/src/abstract-bust-blueprint-v2.ts
import schemaDocument from "../contracts/abstract_bust_v0_2/abstract_bust_blueprint_v0_2.schema.json" with { type: "json" };
import defaultDocument from "../contracts/abstract_bust_v0_2/fixtures/default.json" with { type: "json" };
import neutralDocument from "../contracts/abstract_bust_v0_2/fixtures/neutral.json" with { type: "json" };
import feminineDocument from "../contracts/abstract_bust_v0_2/fixtures/feminine.json" with { type: "json" };
import masculineDocument from "../contracts/abstract_bust_v0_2/fixtures/masculine.json" with { type: "json" };
import youthfulDocument from "../contracts/abstract_bust_v0_2/fixtures/youthful.json" with { type: "json" };
import balancedDocument from "../contracts/abstract_bust_v0_2/fixtures/balanced.json" with { type: "json" };
import matureDocument from "../contracts/abstract_bust_v0_2/fixtures/mature.json" with { type: "json" };
import hairNoneDocument from "../contracts/abstract_bust_v0_2/fixtures/hair_none.json" with { type: "json" };
import hairShortDocument from "../contracts/abstract_bust_v0_2/fixtures/hair_short.json" with { type: "json" };
import hairMediumDocument from "../contracts/abstract_bust_v0_2/fixtures/hair_medium.json" with { type: "json" };
import hairLongDocument from "../contracts/abstract_bust_v0_2/fixtures/hair_long.json" with { type: "json" };
import hairTiedDocument from "../contracts/abstract_bust_v0_2/fixtures/hair_tied.json" with { type: "json" };

export const ABSTRACT_BUST_V2_GENERATOR_VERSION =
  "abstract_bust_v0_2" as const;

export type AbstractBustPresentationV2 =
  | "neutral"
  | "feminine"
  | "masculine";
export type AbstractBustAgeTendencyV2 =
  | "youthful"
  | "balanced"
  | "mature";
export type AbstractBustHairStyleV2 =
  | "none"
  | "short"
  | "medium"
  | "long"
  | "tied";
export type AbstractBustPresetKindV2 =
  | "presentation"
  | "age_tendency"
  | "hair";

export type AbstractBustEditorFieldDescriptorV2 = Readonly<{
  path: string;
  valueType: "number" | "integer" | "enum";
  defaultValue: number | string;
  minimum?: number;
  maximum?: number;
  enumValues?: readonly string[];
}>;

export type AbstractBustBlueprintV2 = {
  generator_version: typeof ABSTRACT_BUST_V2_GENERATOR_VERSION;
  particle_count: number;
  presentation: AbstractBustPresentationV2;
  age_tendency: AbstractBustAgeTendencyV2;
  seed: number;
  head: {
    width: number;
    height: number;
    depth: number;
    roundness: number;
  };
  face: {
    eyes: {
      vertical_position: number;
      spacing: number;
      size: number;
      tilt: number;
      contour_strength: number;
    };
    nose: {
      vertical_position: number;
      width: number;
      length: number;
      prominence: number;
    };
    mouth: {
      vertical_position: number;
      width: number;
      curvature: number;
      contour_strength: number;
    };
    cheeks: {
      width: number;
      vertical_position: number;
      prominence: number;
    };
    jaw: {
      width: number;
      taper: number;
      length: number;
      roundness: number;
    };
  };
  neck: {
    width: number;
    length: number;
  };
  shoulders: {
    width: number;
    slope: number;
  };
  torso: {
    width: number;
    thickness: number;
    length: number;
    taper: number;
  };
  hair: {
    style: AbstractBustHairStyleV2;
    volume: number;
    length: number;
  };
  contour: {
    softening: number;
    asymmetry: number;
  };
};

type JsonObject = Record<string, unknown>;
type SchemaNode = JsonObject & {
  $ref?: string;
  const?: unknown;
  enum?: unknown[];
  type?: string;
  minimum?: number;
  maximum?: number;
  default?: unknown;
  required?: string[];
  properties?: Record<string, SchemaNode>;
};

const schema = schemaDocument as unknown as SchemaNode & {
  $defs: Record<string, SchemaNode>;
};
const defaultBlueprint =
  defaultDocument as unknown as AbstractBustBlueprintV2;
const presentationPresets = {
  neutral: neutralDocument,
  feminine: feminineDocument,
  masculine: masculineDocument,
} as const;
const agePresets = {
  youthful: youthfulDocument,
  balanced: balancedDocument,
  mature: matureDocument,
} as const;
const hairPresets = {
  none: hairNoneDocument,
  short: hairShortDocument,
  medium: hairMediumDocument,
  long: hairLongDocument,
  tied: hairTiedDocument,
} as const;

export const ABSTRACT_BUST_V2_PRESENTATIONS = Object.freeze(
  [
    ...((schema.properties?.presentation?.enum ??
      []) as AbstractBustPresentationV2[]),
  ]
);
export const ABSTRACT_BUST_V2_AGE_TENDENCIES = Object.freeze(
  [
    ...((schema.properties?.age_tendency?.enum ??
      []) as AbstractBustAgeTendencyV2[]),
  ]
);
const hairSchemaReference = schema.properties?.hair?.$ref;
const hairSchemaName = hairSchemaReference?.replace("#/$defs/", "") ?? "";
export const ABSTRACT_BUST_V2_HAIR_STYLES = Object.freeze(
  [
    ...((schema.$defs[hairSchemaName]?.properties?.style?.enum ??
      []) as AbstractBustHairStyleV2[]),
  ]
);

export class AbstractBustBlueprintV2ValidationError extends Error {
  readonly code: string;
  readonly path: string;

  constructor(code: string, path: string, message: string) {
    super(`${code} at ${path}: ${message}`);
    this.name = "AbstractBustBlueprintV2ValidationError";
    this.code = code;
    this.path = path;
  }
}

function isRecord(value: unknown): value is JsonObject {
  return Boolean(value && typeof value === "object" && !Array.isArray(value));
}

function cloneJson<T>(value: T): T {
  return JSON.parse(JSON.stringify(value)) as T;
}

function resolveSchema(value: SchemaNode): SchemaNode {
  if (!value.$ref) return value;
  const prefix = "#/$defs/";
  if (!value.$ref.startsWith(prefix)) {
    throw new Error(
      `Unsupported AbstractBustBlueprintV2 schema reference: ${value.$ref}`
    );
  }
  const resolved = schema.$defs[value.$ref.slice(prefix.length)];
  if (!resolved) {
    throw new Error(
      `Missing AbstractBustBlueprintV2 schema definition: ${value.$ref}`
    );
  }
  return resolved;
}

function collectEditorFields(
  schemaValue: SchemaNode,
  path: readonly string[] = []
): AbstractBustEditorFieldDescriptorV2[] {
  const currentSchema = resolveSchema(schemaValue);
  if (currentSchema.type === "object") {
    return Object.entries(currentSchema.properties ?? {}).flatMap(
      ([fieldName, fieldSchema]) => {
        if (path.length === 0 && fieldName === "generator_version") return [];
        return collectEditorFields(fieldSchema, [...path, fieldName]);
      }
    );
  }

  const defaultValue = currentSchema.default;
  const fieldPath = path.join(".");
  if (currentSchema.enum) {
    if (typeof defaultValue !== "string") {
      throw new Error(`Missing enum editor default at ${fieldPath}`);
    }
    return [
      {
        path: fieldPath,
        valueType: "enum",
        defaultValue,
        enumValues: Object.freeze(currentSchema.enum.map(String)),
      },
    ];
  }
  if (currentSchema.type === "number" || currentSchema.type === "integer") {
    if (
      typeof defaultValue !== "number" ||
      typeof currentSchema.minimum !== "number" ||
      typeof currentSchema.maximum !== "number"
    ) {
      throw new Error(`Incomplete numeric editor schema at ${fieldPath}`);
    }
    return [
      {
        path: fieldPath,
        valueType: currentSchema.type,
        defaultValue,
        minimum: currentSchema.minimum,
        maximum: currentSchema.maximum,
      },
    ];
  }
  return [];
}

export const ABSTRACT_BUST_V2_EDITOR_FIELDS: readonly AbstractBustEditorFieldDescriptorV2[] =
  Object.freeze(
    collectEditorFields(schema).map((field) => Object.freeze(field))
  );

function fail(code: string, path: string, message: string): never {
  throw new AbstractBustBlueprintV2ValidationError(code, path, message);
}

function normalizeValue(
  value: unknown,
  schemaValue: SchemaNode,
  defaultValue: unknown,
  path: string
): unknown {
  const currentSchema = resolveSchema(schemaValue);

  if (
    Object.prototype.hasOwnProperty.call(currentSchema, "const") &&
    (typeof value !== typeof currentSchema.const ||
      value !== currentSchema.const)
  ) {
    fail(
      path === "$.generator_version"
        ? "unknown_generator_version"
        : "invalid_const",
      path,
      `expected ${JSON.stringify(currentSchema.const)}`
    );
  }

  if (
    currentSchema.enum &&
    !currentSchema.enum.some(
      (candidate) =>
        typeof value === typeof candidate && value === candidate
    )
  ) {
    fail("unknown_enum", path, `unsupported value ${JSON.stringify(value)}`);
  }

  if (currentSchema.type === "object") {
    if (!isRecord(value)) fail("invalid_type", path, "expected object");
    const properties = currentSchema.properties;
    if (!properties) {
      throw new Error(`Object schema has no properties at ${path}`);
    }
    const unknownFields = Object.keys(value)
      .filter((key) => !(key in properties))
      .sort();
    if (unknownFields.length) {
      fail(
        "unknown_field",
        `${path}.${unknownFields[0]}`,
        "additional properties are forbidden"
      );
    }
    const requiredFields = new Set(currentSchema.required ?? []);
    const missingRequired = [...requiredFields]
      .filter((key) => !(key in value))
      .sort();
    if (missingRequired.length) {
      fail(
        "missing_required_field",
        `${path}.${missingRequired[0]}`,
        "required field is missing"
      );
    }

    const defaults = isRecord(defaultValue) ? defaultValue : {};
    const normalized: JsonObject = {};
    for (const [fieldName, fieldSchema] of Object.entries(properties)) {
      if (fieldName in value) {
        normalized[fieldName] = normalizeValue(
          value[fieldName],
          fieldSchema,
          defaults[fieldName],
          `${path}.${fieldName}`
        );
        continue;
      }
      if (requiredFields.has(fieldName)) continue;
      const hasFixtureDefault = fieldName in defaults;
      const hasSchemaDefault = Object.prototype.hasOwnProperty.call(
        fieldSchema,
        "default"
      );
      if (!hasFixtureDefault && !hasSchemaDefault) continue;
      const optionalDefault = cloneJson(
        hasFixtureDefault ? defaults[fieldName] : fieldSchema.default
      );
      normalized[fieldName] = normalizeValue(
        optionalDefault,
        fieldSchema,
        optionalDefault,
        `${path}.${fieldName}`
      );
    }
    return normalized;
  }

  if (currentSchema.type === "integer") {
    if (typeof value !== "number" || !Number.isInteger(value)) {
      fail("invalid_type", path, "expected integer");
    }
    let normalized = value;
    if (typeof currentSchema.minimum === "number") {
      normalized = Math.max(currentSchema.minimum, normalized);
    }
    if (typeof currentSchema.maximum === "number") {
      normalized = Math.min(currentSchema.maximum, normalized);
    }
    return normalized;
  }

  if (currentSchema.type === "number") {
    if (typeof value !== "number") {
      fail("invalid_type", path, "expected number");
    }
    if (!Number.isFinite(value)) {
      fail("non_finite_number", path, "number must be finite");
    }
    let normalized = value;
    if (typeof currentSchema.minimum === "number") {
      normalized = Math.max(currentSchema.minimum, normalized);
    }
    if (typeof currentSchema.maximum === "number") {
      normalized = Math.min(currentSchema.maximum, normalized);
    }
    return normalized;
  }

  return value;
}

export function normalizeAbstractBustBlueprintV2(
  value: unknown
): AbstractBustBlueprintV2 {
  let parsed = value;
  if (typeof value === "string") {
    try {
      parsed = JSON.parse(value);
    } catch (error) {
      fail(
        "invalid_json",
        "$",
        error instanceof Error ? error.message : String(error)
      );
    }
  }
  return normalizeValue(
    parsed,
    schema,
    defaultBlueprint,
    "$"
  ) as AbstractBustBlueprintV2;
}

export const parseAbstractBustBlueprintV2 = normalizeAbstractBustBlueprintV2;

export function getDefaultAbstractBustBlueprintV2(): AbstractBustBlueprintV2 {
  return normalizeAbstractBustBlueprintV2(cloneJson(defaultBlueprint));
}

export function getAbstractBustPresetFixtureV2(
  kind: "presentation",
  value: AbstractBustPresentationV2
): AbstractBustBlueprintV2;
export function getAbstractBustPresetFixtureV2(
  kind: "age_tendency",
  value: AbstractBustAgeTendencyV2
): AbstractBustBlueprintV2;
export function getAbstractBustPresetFixtureV2(
  kind: "hair",
  value: AbstractBustHairStyleV2
): AbstractBustBlueprintV2;
export function getAbstractBustPresetFixtureV2(
  kind: AbstractBustPresetKindV2,
  value: string
): AbstractBustBlueprintV2 {
  const fixture =
    kind === "presentation"
      ? presentationPresets[value as AbstractBustPresentationV2]
      : kind === "age_tendency"
        ? agePresets[value as AbstractBustAgeTendencyV2]
        : hairPresets[value as AbstractBustHairStyleV2];
  if (!fixture) {
    fail("unknown_enum", `$.${kind}`, `unsupported value ${JSON.stringify(value)}`);
  }
  return normalizeAbstractBustBlueprintV2(cloneJson(fixture));
}

export function applyAbstractBustPresetV2(
  current: AbstractBustBlueprintV2,
  kind: "presentation",
  value: AbstractBustPresentationV2
): AbstractBustBlueprintV2;
export function applyAbstractBustPresetV2(
  current: AbstractBustBlueprintV2,
  kind: "age_tendency",
  value: AbstractBustAgeTendencyV2
): AbstractBustBlueprintV2;
export function applyAbstractBustPresetV2(
  current: AbstractBustBlueprintV2,
  kind: "hair",
  value: AbstractBustHairStyleV2
): AbstractBustBlueprintV2;
export function applyAbstractBustPresetV2(
  current: AbstractBustBlueprintV2,
  kind: AbstractBustPresetKindV2,
  value: string
): AbstractBustBlueprintV2 {
  const normalizedCurrent = normalizeAbstractBustBlueprintV2(current);
  const preset = getAbstractBustPresetFixtureV2(
    kind as "presentation",
    value as AbstractBustPresentationV2
  );
  return normalizeAbstractBustBlueprintV2({
    ...preset,
    seed: normalizedCurrent.seed,
    particle_count: normalizedCurrent.particle_count,
  });
}
